### PR TITLE
feat: add bip21 addresses for onchain and lightning

### LINKED ISSUE #302: Allow BIP21 strings
Web app should allow and handle (maybe just allow as valid input, don't visually truncate anything) BIP21 `bitcoin:` strings.

- [ ] Allow regular `bitcoin:` BIP21 as valid address input for reverse swaps, `bitcoin:bc1address`
- [ ] Allow `bitcoin:` as valid invoice input for normal swaps if it is a https://bitcoinqr.dev/ with a valid ln invoice
- [ ] `liquidnetwork:` for liquid addresses
- [ ] `lightning:` for lightning invoices

### PR DESCRIPTION
**Pull request by dni**
closes #302

### CODE REVIEW

File: src/utils/invoice.js [hunk 1]
```diff
@@ -47,11 +45,19 @@ export function fetchLnurl(lnurl, amount_sat) {
     });
 }
 
-export function trimLightningPrefix(invoice) {
-    if (invoice.toLowerCase().startsWith(invoicePrefix)) {
-        return invoice.slice(invoicePrefix.length);
+export function trimPrefix(invoice, prefix) {
```
Comment (michael1011) [inline src/utils/invoice.js:48]: Horrible function name. Something like `extractBip21Address` would be better

File: src/utils/invoice.js [hunk 2]
```diff
@@ -47,11 +45,19 @@ export function fetchLnurl(lnurl, amount_sat) {
...
+    if (invoice.toLowerCase().startsWith(prefix)) {
+        const url = new URL(invoice);
+        return url.pathname;
     }
+    return invoice;
+}
 
+export function trimBip21Lightning(invoice) {
```
Comment (michael1011) [inline src/utils/invoice.js:56]: Same here. `extractBip21Invoice`

File: src/utils/invoice.js [hunk 3]
```diff
@@ -47,11 +45,19 @@ export function fetchLnurl(lnurl, amount_sat) {
     });
 }
 
-export function trimLightningPrefix(invoice) {
-    if (invoice.toLowerCase().startsWith(invoicePrefix)) {
-        return invoice.slice(invoicePrefix.length);
+export function trimPrefix(invoice, prefix) {
+    if (invoice.toLowerCase().startsWith(prefix)) {
+        const url = new URL(invoice);
+        return url.pathname;
```
Comment (michael1011) [inline src/utils/invoice.js:51]: And there is no test making sure the URL query parameters of the `invoice` variable are removed

**Review by michael1011 — CHANGES_REQUESTED** [review]

**Review by kilrau — CHANGES_REQUESTED** [review]
Reverse Swap BIP21 works with segwit addresses:
image

But not with nested or legacy:
image
image

image

**Comment by kilrau** [comment]
Reverse swap works, but don't truncate, just leave the whole BIP21 unified string there just like you do for normal swaps.

Also make sure that tests cover it all.

**Review by kilrau — CHANGES_REQUESTED** [review]
Parsing unified BIP21s (https://bitcoinqr.dev/) for normal swaps works fine.

Reverse swaps still several issues:
* `liquidnetwork:` bip21s are not recognized
* `bitcoin:` with legacy P2PKH (1...) or P2WPKH-nested-in-P2SH (3...) still not recognized as I mentioned in my previous review.

**Comment by dni** [comment]
waiting for other prs to resolve first

**Pull request closed without merging**